File: typescript/src/routes/normalize.ts
import express from "express";

const router = express.Router();
router.post("/normalize", (req, res) => {
  const data = req.body.data;

  if (!Array.isArray(data) || data.some((item) => typeof item !== "number")) {
    res.status(422).send({ error: "Input must be an array of numbers." });
    return;
  }

  const max = Math.max(...data);
  const min = Math.min(...data);
  const normalizedData =
    max === min
      ? data.map(() => 0)
      : data.map((num) => (num - min) / (max - min));

  res.send({ normalizedData });
});

export default router;

File: ruby/app/routes/normalize.rb
# frozen_string_literal: true

post '/normalize' do
  data = JSON.parse(request.body.read)['data']
  if !data.is_a?(Array) || data.any? { |item| !item.is_a?(Numeric) }
    halt 422, { error: 'Input must be an array of numbers.' }.to_json
  end

  max = data.max
  min = data.min
  normalized_data = max == min ? Array.new(data.length, 0) : data.map { |num| (num - min).to_f / (max - min) }

  { normalizedData: normalized_data }.to_json
rescue JSON::ParserError
  halt 400, { error: 'Bad request' }.to_json
end
